mmdet/apis/ote/apis/detection/nncf_task.py

```diff
@@ -0,0 +1,211 @@
+}
+
+
+class OTEDetectionNNCFTask(OTEDetectionInferenceTask, IOptimizationTask):
+
+    def __init__(self, task_environment: TaskEnvironment):
+        """"
+        Task for compressing object detection models using NNCF.
+        """
+        super().__init__(task_environment)
+
+        # NNCF part
+        self._compression_ctrl = None
+        nncf_config_path = os.path.join(self._base_dir, "compression_config.json")
+
+        with open(nncf_config_path) as nncf_config_file:
+            common_nncf_config = json.load(nncf_config_file)
+
+        optimization_type = COMPRESSION_MAP[self._hyperparams.nncf_optimization.preset]
+
+        optimization_config = compose_nncf_config(common_nncf_config, [optimization_type])
+        self._config.update(optimization_config)
+
+        # Create and initialize PyTorch model.
+        check_nncf_is_enabled()
+        self._compression_ctrl, self._model = self._load_nncf_model(task_environment.model)
+
+    def _load_model(self, model: ModelEntity):
+        # disable _model_load to load mmdet model and use _load_nncf_model
+        return None
+
+    def _load_nncf_model(self, model: ModelEntity):
+        compression_ctrl = None
+        if model is not None:
+            # If a model has been trained and saved for the task already, create empty model and load weights here
+            buffer = io.BytesIO(model.get_data("weights.pth"))
+            model_data = torch.load(buffer, map_location=torch.device('cpu'))
+
+            model = self._create_model(self._config, from_scratch=True)
+            try:
+                if is_state_nncf(model_data):
+                    compression_ctrl, model = wrap_nncf_model(
+                        model,
+                        self._config,
+                        init_state_dict=model_data,
+                        get_fake_input_func=get_fake_input
+                    )
+                    logger.info("Loaded model weights from Task Environment and wrapped by NNCF")
+                else:
+                    try:
+                        model.load_state_dict(model_data['model'])
+                        logger.info(f"Loaded model weights from Task Environment")
+                        logger.info(f"Model architecture: {self._model_name}")
+                    except BaseException as ex:
+                        raise ValueError("Could not load the saved model. The model file structure is invalid.") \
+                            from ex
+
+                logger.info(f"Loaded model weights from Task Environment")
+                logger.info(f"Model architecture: {self._model_name}")
+            except BaseException as ex:
+                raise ValueError("Could not load the saved model. The model file structure is invalid.") \
+                    from ex
+        else:
+            raise ValueError(f"No trained model in project. NNCF require pretrained weights to compress the model")
+        return compression_ctrl, model
```

Comment:
Can all this stuff be just `_load_model`?

Reply:
No,  _load_model runs before modification config for NNCf, _load_model and _load_nncf_model returns different parameters.

Reply:
I meant: Can we do something like this:
```

def __init__(self):
        self._compression_ctrl = None
        super().__init__()
      
def _load_model(self, model: ModelEntity):
        nncf_config_path = os.path.join(self._base_dir, "compression_config.json")

        with open(nncf_config_path) as nncf_config_file:
            common_nncf_config = json.load(nncf_config_file)

        optimization_type = COMPRESSION_MAP[self._hyperparams.nncf_optimization.preset]

        optimization_config = compose_nncf_config(common_nncf_config, [optimization_type])
        self._config.update(optimization_config)

        # Create and initialize PyTorch model.
        check_nncf_is_enabled()
        model = task_environment.model

        compression_ctrl = None
        if model is not None:
            # If a model has been trained and saved for the task already, create empty model and load weights here
            buffer = io.BytesIO(model.get_data("weights.pth"))
            model_data = torch.load(buffer, map_location=torch.device('cpu'))

            model = self._create_model(self._config, from_scratch=True)
            try:
                if is_state_nncf(model_data):
                    compression_ctrl, model = wrap_nncf_model(
                        model,
                        self._config,
                        init_state_dict=model_data,
                        get_fake_input_func=get_fake_input
                    )
                    logger.info("Loaded model weights from Task Environment and wrapped by NNCF")
                else:
                    try:
                        model.load_state_dict(model_data['model'])
                        logger.info(f"Loaded model weights from Task Environment")
                        logger.info(f"Model architecture: {self._model_name}")
                    except BaseException as ex:
                        raise ValueError("Could not load the saved model. The model file structure is invalid.") \
                            from ex

                logger.info(f"Loaded model weights from Task Environment")
                logger.info(f"Model architecture: {self._model_name}")
            except BaseException as ex:
                raise ValueError("Could not load the saved model. The model file structure is invalid.") \
                    from ex
        else:
            raise ValueError(f"No trained model in project. NNCF require pretrained weights to compress the model")
        self._compression_ctrl = compression_ctrl
        return model
```

Reply:
done